### PR TITLE
[CS-1569] Support perMessageTtl for retry queue

### PR DESCRIPTION
**Pull request by ergurjeet**
**Problem:**
We cannot change `messageTtl` on a queue

**Proposed Solution:**
Don't set  message TTL on queue, set it on per message basis

**Implementation:**
To enable per message TTL on a new exchange, do:
```
{
...
perMessageTtl: true,
}
```
https://github.com/gas-buddy/configured-rabbitmq-client/pull/41/files#diff-9a2ef05db639148d956ff61a1331ad1aR48
- `messageTtl` is for rabbot to set on queue.
- `perMessageTtl` is for this client to set it on per message.

### CODE REVIEW

File: src/exchangeGroups.js [hunk 1]
```diff
@@ -43,16 +43,17 @@ function normalizeExchangeGroup(key, group) {
   normalized.queue = groupFromInput(group.queue, { name: `${exchangeName}.q` });
 
   if (normalized.retries) {
-    normalized.retryExchange = groupFromInput(group.retryExchange, { name: `${exchangeName}.retry`, persistent });
+    const [suffix, ttlKey] = group.perMessageTtl === true ? ['.nottl', 'perMessageTtl'] : ['', 'messageTtl'];
```
Comment (djMax) [inline src/exchangeGroups.js]: I don't love these suffixes... Why are they necessary?

File: src/exchangeGroups.js [hunk 2]
```diff
@@ -91,9 +92,7 @@ export function rabbotConfigFromExchangeGroups(exchangeGroups) {
     if (g.retryExchange) {
       const retryExchange = Object.assign({}, exchangeDefaults, g.retryExchange);
       const retryQueue = Object.assign({}, queueDefaults, g.retryQueue);
-      retryQueue.messageTtl = g.retryDelay;
```
Comment (djMax) [inline src/exchangeGroups.js:94]: This is a backwards incompatible change. I think it would be better to simply make it an option on the exchange group and use that to publish to the retryQueue as appropriate.

File: tests/test_config.js [hunk 3]
```diff
@@ -128,30 +128,42 @@ tap.test('test routing key reuse', async (t) => {
 
 tap.test('test delivery_mode pass thru', async (t) => {
   const retryCount = 2;
+  const retryDelay = 380;
   const mq = new RabbotClient(ctx, configWithExchangeGroups({
     persistent: {
       retries: retryCount,
-      retryDelay: 100,
```
Comment (djMax) [inline tests/test_config.js:134]: Why is this removed?

File: src/exchangeGroups.js (hunk 1 above)
Comment (ergurjeet) [reply src/exchangeGroups.js]: Existing retry queues have `messageTtl` (you cannot reset that). You need a different name for new queues.

File: src/exchangeGroups.js (hunk 2 above)
Comment (ergurjeet) [reply src/exchangeGroups.js:94]: Not removed. It's moved up the hierarch a bit https://github.com/gas-buddy/configured-rabbitmq-client/pull/41/files#diff-9a2ef05db639148d956ff61a1331ad1aR49. Everything should work normal if you don't set perMessageTtl: true. If that value is set, then it will create a new queue without messageTtl and start passing expiresAfter per message https://github.com/gas-buddy/configured-rabbitmq-client/pull/41/files#diff-1fdf421c05c1140f6d71444ea2b27638R297

File: tests/test_config.js (hunk 3 above)
Comment (ergurjeet) [reply tests/test_config.js:134]: Not removed. Set a const https://github.com/gas-buddy/configured-rabbitmq-client/pull/41/files#diff-b0c355da8db98fa65b6b974b5977053dR131 so it can be used to assert claims here https://github.com/gas-buddy/configured-rabbitmq-client/pull/41/files#diff-b0c355da8db98fa65b6b974b5977053dR151

File: src/exchangeGroups.js (hunk 1 above)
Comment (ergurjeet) [reply src/exchangeGroups.js]: I agree that once this setting is switched on/off, it will leave an orphan queue (only one irrespective of how many times you toggle it). However, no one will switch it on a regular basis. I would like to test it on a few exchanges before setting it as default.

File: src/exchangeGroups.js (hunk 2 above)
Comment (ergurjeet) [reply src/exchangeGroups.js:94]: We can set `retryDelay` on an exchange group. These changes will respect that setting

File: src/exchangeGroups.js (hunk 1 above)
Comment (djMax) [reply src/exchangeGroups.js]: But if you DON'T change the built in TTL on an exchange, the perMessageTtl will work just fine, right?

File: src/index.js [hunk 4]
```diff
@@ -294,6 +294,7 @@ the MQ_MAKE_EXCHANGES environment variable and restart.
                 routingKey: message.fields.routingKey,
                 correlationId: message.properties.correlationId,
                 timestamp: message.properties.timestamp,
+                expiresAfter: exchangeGroup.retryQueue.perMessageTtl || undefined,
```
Comment (djMax) [inline src/index.js:298]: This means the "per message ttl" is not in fact per message, right? It's just using a different way to execute the configured ttl?

File: src/exchangeGroups.js (hunk 1 above)
Comment (ergurjeet) [reply src/exchangeGroups.js]: It will take the lower of two. You can reduce it, but can't increase it beyond the limit specified in a queue

File: src/index.js (hunk 4 above)
Comment (ergurjeet) [reply src/index.js:298]: Yes it's a different way to use the configured ttl, however, instead of setting it on a queue (that we cannot change later), this will set on individual message (that we can change).

File: src/exchangeGroups.js (hunk 1 above)
Comment (ergurjeet) [reply src/exchangeGroups.js]: `.nottl` could be `v2`, however, `.nottl` reflects the intention

**Comment by ergurjeet** [comment]
@djMax - Checking again if this is something we should pursue

File: src/exchangeGroups.js (hunk 1 above)
Comment (djMax) [reply src/exchangeGroups.js]: I don't generally like this "additional magic" names. I would prefer that the new options are just that - new options. So if someone wants to use them, they have to change the name of the exchange on their own, not have this module do these opinionated automatic things.

**Comment by ergurjeet** [comment]
@djMax I have made the changes as per your last comment.

File: src/exchangeGroups.js (hunk 2 above)
Comment (djMax) [inline src/exchangeGroups.js:94]: don't you need to keep this so that existing retry queues don't mismatch?

Comment (djMax) [reply src/exchangeGroups.js:94]: Oh - I see, it's managed up above, right?

Comment (ergurjeet) [reply src/exchangeGroups.js:94]: Yes

Comment (djMax) [reply src/exchangeGroups.js:94]: Difference is now it's controlled by normalized.retries instead of g.retryExchange. I guess those are the same object?

Comment (ergurjeet) [reply src/exchangeGroups.js:94]: From same config property. Now it's decided at the time of normalization

Comment (ergurjeet) [reply src/exchangeGroups.js:94]: g.retryQueue and normalized.retryQueue are same object
